### PR TITLE
add tests for bpr_max and NCE losses

### PR DESCRIPTION
**Pull request by denkuzin**
# Description

## Type of change

- [x] Feature
- [ ] Fix
- [ ] Release

## Checklist

- [x] I have followed the [CONTRIBUTING](https://github.com/criteo/deepr/blob/master/docs/CONTRIBUTING.rst) guidelines.
- [ ] I have updated the [CHANGELOG](https://github.com/criteo/deepr/blob/master/docs/CHANGELOG.rst).

### CODE REVIEW

File: deepr/layers/core.py [hunk 1]
```diff
@@ -160,3 +160,17 @@ def forward(self, tensors, mode: str = None):
...
+class Softmax(base.Layer):
+    """Apply softmax to the last dimension of tensor with filtering masked values"""
+
+    def __init__(self, n_in: int = 2, n_out=1, **kwargs):
+        super().__init__(n_in=n_in, n_out=n_out, **kwargs)
+
+    def forward(self, tensors, mode: str = None):
+        """Forward method of the layer"""
+        tensor, mask = tensors
+        tensor_exp = tf.exp(tensor - tf.reduce_max(tensor, axis=-1, keepdims=True))
```
Comment (guillaumegenthial) [inline deepr/layers/core.py:174]: the reduce max is here for numerical stability right?
Also, shouldn't it be also multiplied by the mask here?

File: deepr/layers/core.py [hunk 2]
```diff
@@ -160,3 +160,17 @@ def forward(self, tensors, mode: str = None):
...
+        sum_tensor_exp = tf.reduce_sum(tf.multiply(tensor_exp, mask), axis=-1, keepdims=True)
```
Comment (guillaumegenthial) [inline deepr/layers/core.py:175]: mask is bool?

File: deepr/layers/core.py [hunk 3]
```diff
@@ -160,3 +160,17 @@ def forward(self, tensors, mode: str = None):
             use_bias=self.use_bias,
             **self._kwargs
         )
+
+
+class Softmax(base.Layer):
+    """Apply softmax to the last dimension of tensor with filtering masked values"""
+
+    def __init__(self, n_in: int = 2, n_out=1, **kwargs):
+        super().__init__(n_in=n_in, n_out=n_out, **kwargs)
```
Comment (guillaumegenthial) [inline deepr/layers/core.py:169]: maybe add a test for that class (masking seems important to get right)

File: deepr/layers/losses.py [hunk 4]
```diff
@@ -223,8 +219,8 @@ def forward(self, tensors, mode: str = None):
 
         losses = tf.add(true_losses, sampled_losses)
         # One loss per event, average of scores : (batch, num_events)
-        event_scores = WeightedAverage()((losses, tf.to_float(mask)))
+        event_scores = WeightedAverage()((losses, mask))
```
Comment (guillaumegenthial) [inline deepr/layers/losses.py:222]: ah, so you expect the mask to already be a float tensor?
it would maybe be safer to expect a bool, especially because we also need it

File: tests/unit/layers/test_layers_losses.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
+"""Tests for layers.bpr"""
+
+import numpy as np
+import tensorflow as tf
+
+import deepr as dpr
+
+
+def test_layers_bpr():
```
Comment (guillaumegenthial) [inline tests/unit/layers/test_layers_losses.py:9]: I see that you made on file with all the losses -> hence one file for test.
maybe more readable to split each class in its own file but just a nit

File: tests/unit/layers/test_layers_losses.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+    # Compare with Tensorflow Value
+    inputs = {key: tf.constant(val, dtype=tf.bool if key == "mask" else tf.float32) for key, val in inputs.items()}
+    result = nce_layer(inputs)
+    with tf.Session() as sess:
+        result_eval = sess.run(result["loss"])
+        np.testing.assert_allclose(result_eval, expected, 1e-4)
+
+
+def test_layers_bpr_max():
+    """Compare BPR Max `Layer` output with a dummy NumPy implementation"""
```
Comment (guillaumegenthial) [inline tests/unit/layers/test_layers_losses.py:138]: I'm not reading this in details, unless you think it's better (I assume if TF and naive NP agree, then we're good)

File: deepr/layers/losses.py (hunk 4 above)
Comment (denkuzin) [reply deepr/layers/losses.py:222]: yes, I do cast upper in the code

File: deepr/layers/core.py (hunk 2 above)
Comment (denkuzin) [reply deepr/layers/core.py:175]: nope, here mask is float

File: deepr/layers/core.py (hunk 3 above)
Comment (denkuzin) [reply deepr/layers/core.py:169]: yes, good point

File: tests/unit/layers/test_layers_losses.py (hunk 5 above)
Comment (denkuzin) [reply tests/unit/layers/test_layers_losses.py:9]: Yes, why not

Comment (denkuzin) [reply tests/unit/layers/test_layers_losses.py]: splitted

File: deepr/layers/core.py (hunk 3 above)
Comment (denkuzin) [reply deepr/layers/core.py:169]: added tests
